testkit/src/mill/testkit/BspServerUtil.scala

```diff
@@ -0,0 +1,184 @@
+package mill.testkit
+
+import ch.epfl.scala.{bsp4j => b}
+import com.google.gson.{Gson, GsonBuilder}
+import coursier.cache.CacheDefaults
+import mill.api.BuildInfo
+import mill.bsp.Constants
+import org.eclipse.{lsp4j => l}
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{ExecutorService, Executors, ThreadFactory}
+
+import scala.jdk.CollectionConverters._
+import scala.reflect.ClassTag
+
+object BspServerUtil {
+
+  val updateFixtures = false
+
+  lazy val bsp4jVersion: String = sys.props.getOrElse("BSP4J_VERSION", ???)
+
+  trait DummyBuildClient extends b.BuildClient {
+    def onBuildLogMessage(params: b.LogMessageParams): Unit = ()
+    def onBuildPublishDiagnostics(params: b.PublishDiagnosticsParams): Unit = ()
+    def onBuildShowMessage(params: b.ShowMessageParams): Unit = ()
+    def onBuildTargetDidChange(params: b.DidChangeBuildTarget): Unit = ()
+    def onBuildTaskFinish(params: b.TaskFinishParams): Unit = ()
+    def onBuildTaskProgress(params: b.TaskProgressParams): Unit = ()
+    def onBuildTaskStart(params: b.TaskStartParams): Unit = ()
+    def onRunPrintStderr(params: b.PrintParams): Unit = ()
+    def onRunPrintStdout(params: b.PrintParams): Unit = ()
+  }
+
+  lazy val dummyBuildClient: b.BuildClient = new DummyBuildClient {}
+
+  val gson: Gson = new GsonBuilder().setPrettyPrinting().create()
+  def compareWithGsonFixture[T: ClassTag](
+      value: T,
+      fixturePath: os.Path,
+      replaceAll: Seq[(String, String)] = Nil
+  ): Unit = {
+
+    def doReplaceAll(input: String, inverse: Boolean = false): String =
+      replaceAll.foldLeft(input) {
+        case (input0, (from0, to0)) =>
+          val (from, to) = if (inverse) (to0, from0) else (from0, to0)
+          input0.replace(from, to)
+      }
+
+    val exists = os.exists(fixturePath)
```

Comment:
I assume `exists -== false` only when generating fixtures for the first time; this should go in a comment or something because it's not obvious